Fix usage of millis() to avoid overflow

MJ666 opened the pull request:
This PR fixes issues with causes an overflow of the casted millis() usage every ~64 seconds. This overflow will potentially create problems in some situations.

MJ666 commented:
Rebased to master.

dgatf commented:
Maybe I'm missing something but I believe the actual code works as expected when casting millis() to uint16_t even when it overflows

See this example, which overflows the second variable when casting and still produce the same result whether is casted or not:

```
void setup() {
  Serial.begin(115200);
  Serial.println();
  uint32_t ts1 = 0xFFFFFFF0;
  uint32_t ts2 = ts1 + 0xFF; 
  uint16_t ts = ts1;
  Serial.println(ts1, HEX);
  Serial.println(ts2, HEX);
  Serial.println(ts2 - ts1);
  Serial.println((uint16_t)ts2 - ts1);
}

void loop() {
}
```

The issue of this PR is that it increases by 238 bytes the flash size and I want to keep atmegas with 32k of flash. Although still fits, DEBUG won't feasible on the pro mini

_Master branch
RAM:   [====      ]  44.6% (used 913 bytes from 2048 bytes)
Flash: [==========]  96.9% (used 29766 bytes from 30720 bytes)_

_Millis
RAM:   [====      ]  44.7% (used 915 bytes from 2048 bytes)
Flash: [==========]  97.7% (used 30004 bytes from 30720 bytes)_

Could you provide an example where the current code can produce a wrong result

MJ666 commented:
I have added back casts if possible and rebased the code to actual master again. Here are my results compiling for the Arduino Nano. The difference is now 152 bytes increase only. This is 86 bytes of saving from initial pull request.

Modified code with casts where possible:
Der Sketch verwendet 29918 Bytes (97%) des Programmspeicherplatzes. Das Maximum sind 30720 Bytes.
Globale Variablen verwenden 915 Bytes (44%) des dynamischen Speichers, 1133 Bytes für lokale Variablen verbleiben. Das Maximum sind 2048 Bytes.

Orginal code:
Der Sketch verwendet 29766 Bytes (96%) des Programmspeicherplatzes. Das Maximum sind 30720 Bytes.
Globale Variablen verwenden 913 Bytes (44%) des dynamischen Speichers, 1135 Bytes für lokale Variablen verbleiben. Das Maximum sind 2048 Bytes.

dgatf commented:
Sorry, I didn't explain myself correctly. Also the timestamp variables can stay as uint16. The only change needed are the brackets, which are correct now

MJ666 commented:
After some more review this should fix he problems. found a nice article here wich also will work for the the casted millis() values in the same way.

https://www.norwegiancreations.com/2018/10/arduino-tutorial-avoiding-the-overflow-issue-when-using-millis-and-micros/

dgatf commented:
> I guess this is what you had in mind? Found a nice article here wich also will work for the the casted millis() values in the same way.
> 
> https://www.norwegiancreations.com/2018/10/arduino-tutorial-avoiding-the-overflow-issue-when-using-millis-and-micros/
> 
> Thinking more about this likely your original code should work as intended. Now we calculate with the full resolution actual millis() value with the casted stored millis() value and cast the result. This likely also causes wrong calculation again?

Finally found that `(uint16_t)millis() - ts` was failing only for ARM Cortex and not ATmega.

Changed to `(uint16_t)(millis() - ts)`

Fixed with ffb0ed3b82eb31d2e03882600988e590c5ea60fb to avoid rebasing again

Thanks!
